rfc/007-workload-secrets.md

```diff
@@ -0,0 +1,108 @@
+# RFC 007: Workload secrets
+
+## Objective
+
+Provide Contrast workloads with secret key material, such that:
+
+1. Workloads can use key material to derive additional secrets autonomously.
+2. The key material is stable across workload upgrades.
+3. The key material is stable across manifest generations.
+4. The key material is stable across Contrast upgrades.
+
+## Problem statement
+
+Due to the verification mechanism, Contrast workloads are naturally stateless.
+Unless the container images are encrypted, there is no secure way to embed a persistent secret.
+
+However, many applications need persistence of some sort.
+Since Contrast workloads usually process sensitive data, the data they persist is often going to be sensitive, too.
+As of today, the only option for persistence is a trusted external service that validates mesh certificates, presenting a bootstrapping problem that Contrast is designed to avoid.
+If workloads somehow had access to a stable secret, they could encrypt their sensitive data before sending it to an untrusted storage provider.
+
+A concrete use case would be to host a [Hashicorp Vault] (or [OpenBao]) instance.
+Vault has a concept of [storage sealing] to protect its secret store.
+In a nutshell, Vault encrypts its secrets with a sealing key.
+The Vault process starts in sealed mode, unable to access its own persistence.
+After being provisioned with the unsealing key it can resume normal operation.
+When running Vault as a Contrast workload, access to a persistent secret would allow automatic unsealing without operator intervention.
+
+[Hashicorp Vault]: https://www.hashicorp.com/products/vault
+[OpenBao]: https://openbao.org/
+[storage sealing]: https://developer.hashicorp.com/vault/docs/concepts/seal
+
+## Proposal
+
+### Secret source
+
+Workload secrets would need to be either persisted or deterministically derived.
+The only secure persistence option for the Coordinator is encryption with a key derived from the secret seed.
```

Comment:
I think we should reference the recovery RFC here, since this is where this "secret seed" comes from, right?

Reply:
Done